freqtrade/strategy/interface.py

```diff
@@ -303,6 +303,22 @@ def custom_sell(self, pair: str, trade: Trade, current_time: datetime, current_r
         """
         return None
 
+    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
+                            proposed_stake: float, min_stake: float, max_stake: float,
+                            **kwargs) -> float:
```

Comment:
i'd THINK some information about the current bot's trade(s) state might be useful if it's accessible easily (currently open trades, max_open_trades, current stake in trades [used stake]) - as proposed by @epigramx in #5076.


while the information is available _in theory_ - using `Trade.xxxx()` will make it more complex for the user than providing it as argument - unless there's a big overhead (which there might be for backtesting - eventually (!)).

Using `Trade.xxx()` in backtesting is (while in theory possible) a bit dangerous at the moment, as not all available methods are simulated for backtesting - and as such may return wrong / no results - and it's not documented as "possible for backtesting" either (mainly because of that).
By providing the info to the function we ensure that only allowed methods are used.

(convince me differently...)

Reply:
I was about to say I don't mind since I would use the Trade class etc., but you make a good point if it can't be used for backtesting since it might be a bad idea to use the method at all (in practice) if it can't be backtested. 
So I would say it should be able to replace in full the current methods (which requires slot availability) and there is very little reason not to because the performance impact should be low because trade entries are rare.

Reply:
yes - speed.

I've experimented with this - and by using a Trade-object instead of a LocalTrade - you get a slowdown of around 5-10x (don't remember the exact numbers - and they depend on the strategy as well).
Creating the LocalTrade class was NOT fun - and i only did that due to the necessity - not because "i feel it's a great way to do things".

I've not even created a session - simply using the Trade object makes everything slow (because of the way sqlalchemy works when accessing it's dynamic attributes). 
Profiling it points to some methods 5-10 levels into the sqlalchemy callstack (when using `trade.xxx`) - and it's that many nested function-calls that it's simply not possible to do it fast with an sqlalchemy "enabled" object.

It's not a problem for real trading - where you only access the same attribute a few times per iteration - but in backtesting - where you'll want to do as many candles as possible in the shortest amount of time - this has a quite huge impact.

Reply:
Almost sounds like sqlalchemy was overkill for such a simple database, and a plain object with explicit save method would both provide feature parity and be as fast as needed in all contexts, without hardly being an inconvenience considering how rarely fields are updated and how non-complicated updates are.

Reply:
An in-memory sqlite is also extremely slow compared to a simple in-memory list. Backtesting doesn't require its robustness anyway since if it is stops all data is expected to be lost anyway. It seems clearly better to hold that kind of basic data (like "number of slots used") into basic python primitives during backtesting.

Reply:
Maybe so. But if sqlalchemy wasnt in a mix - we would essentially have a LocalTrade that can save on request, and in backtesting there would be no saving so no sqlite. But most importantly it would be same code used in both modes, not having to worry about overhead, not introducing any complexities.. It would be extremely useful if we could reliably get current trade info in backtesting. It would open new approaches in strategy development.

Reply:
First of all - in `populate_*` methods, you'll never receive any reliable trade data - simply because they run "all at once" - and the trade-simulation starts later.

For the rest - sure - but if you look hard enough you'll notice that the possibility exists - maybe in a slightly limited form (only certain filters are possible).
it's however not documented on purpose - as i'm not sure it'll stay (i hope it doesn't - but didn't have time to look into a proper implementation) - and by not having it exposed through documentation - it'll not matter if it's removed later on).

Reply:
rokups explained to me in discord that wallet_size  from the current patch and  max_open_trades from the config can be used to reliably simulate 'unlimited' (and I guess variations of it) in backtesting. That means the patch might be already enough for anyone that would need 'open trades & total trades & stake available' data during backtesting. I guess a backtesting-available Trade class is not required for that purpose but it might be always useful for ease of use of functions like this.